### PR TITLE
Replace assert(...) by macros from invariant.h [blocks: #6749]

### PR DESCRIPTION
**Pull request by tautschnig**
While new/modified code shouldn't use assert(...) anyway, we had several
existing uses of assert. Once we remove nonstd/optional.h, we will no
longer implicitly include cassert, implying that a lot of #include
<cassert> changes will be necessary. Instead of adding these, go all the
way and fix the undesired uses of assert.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/cpp/cpp_typecheck_expr.cpp [hunk 1]
```diff
@@ -1712,12 +1713,12 @@ void cpp_typecheckt::typecheck_side_effect_function_call(
...
   if(expr.type().id()==ID_constructor)
   {
-    assert(expr.function().id() == ID_symbol);
+    PRECONDITION(expr.function().id() == ID_symbol);
 
     const code_typet::parameterst &parameters=
       to_code_type(expr.function().type()).parameters();
 
-    assert(parameters.size()>=1);
+    DATA_INVARIANT(parameters.size() >= 1, "parameters expected");
```
Comment (thomasspriggs) [inline src/cpp/cpp_typecheck_expr.cpp]: Suggestion - `!parameters.empty()`

File: src/cpp/cpp_typecheck_resolve.cpp [hunk 2]
```diff
@@ -1099,8 +1102,10 @@ struct_tag_typet cpp_typecheck_resolvet::disambiguate_template_classes(
       cpp_declaration.template_type());
 
     // iterate over template instance
-    assert(full_template_args_tc.arguments().size()==
-           partial_specialization_args.arguments().size());
+    DATA_INVARIANT(
+      full_template_args_tc.arguments().size() ==
+        partial_specialization_args.arguments().size(),
+      "argument numbers must match");
```
Comment (thomasspriggs) [inline src/cpp/cpp_typecheck_resolve.cpp]: Suggestion - "Number of arguments should match."
Your wording suggests that each argument has a number which is being matched.

File: src/goto-instrument/goto_program2code.cpp [hunk 3]
```diff
@@ -1250,8 +1252,10 @@ goto_programt::const_targett goto_program2codet::convert_start_thread(
   {
     goto_programt::const_targett this_end=next;
     ++this_end;
-    assert(this_end->is_end_thread());
-    assert(thread_start->location_number > this_end->location_number);
+    DATA_INVARIANT(this_end->is_end_thread(), "should be end-of-thread");
+    DATA_INVARIANT(
+      thread_start->location_number > this_end->location_number,
+      "monotone location numbers");
```
Comment (thomasspriggs) [inline src/goto-instrument/goto_program2code.cpp]: "monotonic"? The musical definition of "monotone" comes to my mind first when reading this. Perhaps a more precise definition of what this is checking is - "Start of new thread should be after the end of existing thread."

File: src/goto-instrument/goto_program2code.cpp [hunk 4]
```diff
@@ -1916,8 +1928,9 @@ void goto_program2codet::cleanup_expr(exprt &expr, bool no_typecast)
          typedef_names.find(typedef_str)==typedef_names.end())
         expr.type().remove(ID_C_typedef);
 
-      assert(expr.type().id()!=ID_union &&
-             expr.type().id()!=ID_struct);
+      DATA_INVARIANT(
+        expr.type().id() != ID_union && expr.type().id() != ID_struct,
+        "struct or union");
```
Comment (thomasspriggs) [inline src/goto-instrument/goto_program2code.cpp]: The condition being checked is actually the inverse - "Type being cast to is not a struct or union."

File: src/goto-instrument/goto_program2code.cpp [hunk 5]
```diff
@@ -1916,8 +1928,9 @@ void goto_program2codet::cleanup_expr(exprt &expr, bool no_typecast)
          typedef_names.find(typedef_str)==typedef_names.end())
         expr.type().remove(ID_C_typedef);
 
-      assert(expr.type().id()!=ID_union &&
-             expr.type().id()!=ID_struct);
+      DATA_INVARIANT(
+        expr.type().id() != ID_union && expr.type().id() != ID_struct,
```
Comment (thomasspriggs) [inline src/goto-instrument/goto_program2code.cpp]: Suggestion might be cleaner to express this as `!can_cast_type<struct_union_typet>(expr.type())`

**Review by thomasspriggs — APPROVED** [review]

File: src/cpp/cpp_instantiate_template.cpp [hunk 6]
```diff
@@ -150,7 +150,7 @@ const symbolt &cpp_typecheckt::class_template_symbol(
   const cpp_template_args_tct &full_template_args)
 {
   // we should never get 'unassigned' here
```
Comment (NlightNFotis) [inline src/cpp/cpp_instantiate_template.cpp]: Can these comments also be removed? They duplicate the intent of the programmatic check.

**Review by NlightNFotis — APPROVED** [review]

File: src/cpp/cpp_instantiate_template.cpp (hunk 6 above)
Comment (tautschnig) [reply src/cpp/cpp_instantiate_template.cpp]: Done!

**Review by peterschrammel — APPROVED** [review]

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 7]
```diff
@@ -765,8 +765,8 @@ void java_bytecode_convert_classt::convert(
     if(s_it!=symbol_table.symbols.end())
       symbol_table.erase(s_it); // erase, we stubbed it
 
-    if(symbol_table.add(new_symbol))
-      assert(false && "failed to add static field symbol");
+    const bool failed = symbol_table.add(new_symbol);
+    INVARIANT(!failed, "failed to add static field symbol");
```
Comment (martin-cs) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: Personally I would `CHECK_RETURN`

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 8]
```diff
@@ -1433,18 +1434,18 @@ java_bytecode_convert_methodt::convert_instructions(const methodt &method)
...
       PRECONDITION(op.empty() && results.empty());
-      assert(!jsr_ret_targets.empty());
+      PRECONDITION(!jsr_ret_targets.empty());
       c = convert_ret(
         jsr_ret_targets, arg0, i_it->source_location, i_it->address);
     }
     else if(bytecode == BC_iconst_m1)
     {
-      assert(results.size()==1);
+      PRECONDITION(results.size() == 1);
```
Comment (martin-cs) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: `CHECK_RETURN`?

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 9]
```diff
@@ -1433,18 +1434,18 @@ java_bytecode_convert_methodt::convert_instructions(const methodt &method)
...
       results[0]=from_integer(-1, java_int_type());
     }
     else if(bytecode == patternt("?const_?"))
     {
-      assert(results.size()==1);
+      PRECONDITION(results.size() == 1);
```
Comment (martin-cs) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: Again?

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 10]
```diff
@@ -1742,7 +1743,7 @@ java_bytecode_convert_methodt::convert_instructions(const methodt &method)
         numeric_cast_v<std::size_t>(to_constant_expr(arg1));
 
       op=pop(dimension);
-      assert(results.size()==1);
+      PRECONDITION(results.size() == 1);
```
Comment (martin-cs) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: Also

File: src/ansi-c/parser_static.inc [hunk 11]
```diff
@@ -221,9 +221,9 @@ static void make_subtype(typet &dest, typet &src)
         break;
       }
       else if(p->is_nil())
-        assert(false);
+        UNREACHABLE;
```
Comment (martin-cs) [inline src/ansi-c/parser_static.inc:224]: A rare annotation but seems right in this case!

**Review by martin-cs — APPROVED** [review]

**Pull request merged**